docs/developer-guide/ksqldb-clients/contributing.md

```diff
@@ -0,0 +1,112 @@
+
+Array, map, and struct types may be recursively nested within each other, and may contain any of the other types as well.
+
+Users of the client may wish to receive result rows in different ways:
+- Do not block. Rather, perform an action asynchronously on each new row as it arrives
+- Block until the next row arrives
+- Block until all rows have arrived. This only makes sense in the context of pull queries and push queries with `LIMIT` clauses. 
+
+Details of client interfaces to support these methods will likely vary by language. For example, the [Java client](java-client.md)
+uses [Reactive Streams](http://www.reactive-streams.org/) for its asynchronous streaming interface. 
+
+### Terminating push queries ###
+
+Requests to terminate push queries should be sent to the [`/close-query` server endpoint](TODO).
+The client should expose an interface that allows users to pass in a push query ID to be passed to this endpoint.
+
+### Inserting new rows of data into existing ksqlDB streams ###
+
+ksqlDB supports inserting new rows of data into existing ksqlDB streams via the [`/inserts-stream` server endpoint](TODO).
+If the server successfully validates the target stream name from a request made to the `/inserts-stream` endpoint,
+the client may then write JSON-serialized rows of data to the connection, and the server will insert each
+asynchronously and write an ack containing the relevant sequence number to the response.
+
+Because the insertions happen asynchronously on the server side, it makes sense for the client to expose APIs
+for inserting one row at a time, but inserting batches of rows is less useful due to the lack of transactional guarantees.
+
+The client should provide a convenient way for users to create rows of data to insert.
+The Java client's [KsqlObject](TODO) is one example, though details will likely vary by language.
+
+### Creation of new streams, tables, and persistent queries ###
+
+These DDL/DML requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
+
+### Terminating persistent queries ###
+
+These requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
+
+### Listing existing streams, tables, topics, and queries ###
+
+These admin operation requests should be sent to the [`/ksql` server endpoint](../ksqldb-rest-api/ksql-endpoint.md).
+
+Configurations
+--------------
+
+The client of course needs to expose options for specifying the address of the ksqlDB server to connect to.
+Additional options important to expose include: 
+- Support for TLS-enabled ksqlDB servers
+- Support for mutual-TLS-enabled ksqlDB servers
+- Support for ksqlDB servers with HTTP basic authentication enabled
+
+It may also be nice to support configuring multiple ksqlDB server addresses, so the client may route requests
+to a different server if one is down.
+
+Process
+-------
+
+Thanks for your interest in contributing a client!
```

Comment:
Might want to move this to the top.

Reply:
Wasn't sure how to move the entire sentence to the top without clashing with the current intro, but I did add a bit of thanks to the top as well. LMK if you had something else in mind.

Reply:
I don't think we should assume anyone writing a client will contribute it back. I'd have the 'how to write a client' section at the top (as it currently is), and then have a 'how to contribute your client to the project' section below.  We should support people just wanting to link their public client repo, as well as those wanting to submit the code to our repo.

The latter has challenges around how we compile such a multi-language project.  The Kafka client's team are probably good people to talk to about this. @edenhill or @ryanatconfluent may be good people to talk to about this.

Reply:
That's a really good point, I hadn't thought about the logistics of compiling a multi-language project. Does it make sense to have clients from other languages in the main ksql repository, or should they be separate repositories?